### DISCUSSION TITLE
Bard field "not working" (still in array format) in custom Tag

### DISCUSSION THREAD

**Discussion by hanneshoepfnerio** [statamic/cms#4579] [Q&A]
Hi, I'm having some issues with bard field content when using a custom tag.
In my custom tag I'm extending  Statamic\Tags\Tags; but I've also tried to use Collection
```class MyCustomTag extends Tags
{
    public function index()
    {
        $entries = Entry::query()
            ->where("collection", "name")
            ->get()
            ->sortBy("date")
            ->where("date", ">=", date("Y-m-d"));

        if ($limit = $this->params->int("limit")) {
            $entries = $entries->take($limit);
        }

   
        return $entries;
    }
```
I'm using it like this
```
{{MyCustomTag}}
  {{MyBardField}} // this is still in the array format magically in e.g <p> tags 
{{/MyCustomTag}}
```
From the docs im guessing that its related to the Bard Augmentor and ProseMirror Rendering, but i don't know if i need to add this to my rendering method, or if its easier to just extend a Class that already works with the bard field.
Thanks for any input on this topic

**Comment by duncanmcclean** [comment]
What about if you return from the tag like this?

```php
return $entries->map(function ($entry) {
    return $entry->toAugmentedArray();
});
```

Does that return what you're looking for?

(fyi: to do codeblocks you need to use three backticks at start and end, not just one)

**Reply by hanneshoepfnerio** [reply]
Hi Duncan, thank you for the reply. I fixed the backticks –
Your suggestion does not seem to work. `$entry->toAugmentedArray();`is not defined, so the method is not available.
I guess its because its just an array and not an Entry type.
Here is the structure of $entries:

```
Statamic\Entries\EntryCollection {#3579 ▼
  #items: array:4 [▼
    6 => array:8 [▼
      "name" => "some name"
      "bardfield" => array:1 [▼
        0 => array:2 [▼
          "type" => "paragraph"
          "content" => array:1 [▶]
        ]
      ]
      "name" => "some name"
      "date" => "2021-10-30 17:37"
      "type" => "mytype"
      "enabled" => true
      "links" => array:2 [▶]
    ]
    0 => array:8 [▶]
    1 => array:8 [▶]
    5 => array:8 [▶]
  ]
  ```
  Maybe you have another idea
  Thank you

**Reply by hanneshoepfnerio** [reply]
Hi, the more I look into it – I think that this is the problem.
Usually the bard field is within an Entry object which is augmentable and resolves the bard field nicely.
In my custom tag im just collection the "raw" array of something containing the bard field. So Entry Collection is not containing Entries, just the array elements itself.
So how can i apply the Entry augmentation which resolves the bard field directly to the value itself?
Or would it also be possible to wrap my array data in an Entry, even though it does not have an ID and all those properties?
I'm happy for input! Tanks

**Reply by edalzell** [reply]
Let's see the output of this:

```php
        $entries = Entry::query()
            ->where("collection", "name")
            ->get()
```

**Comment by hanneshoepfnerio** [comment]
Hi, yes – thats what I also looked into.
Here is the print of the above query:
```
Statamic\Entries\EntryCollection {#3785 ▼
  #items: array:4 [▼
    0 => Statamic\Entries\Entry {#3612 ▼
      #id: "b6fb290d-df09-4f88-a4f2-c32715cd05f9"
      #collection: "col"
      #blueprint: "blue"
      #date: null
      #locale: "default"
      #localizations: null
      #afterSaveCallbacks: []
      #withEvents: true
      #slug: "slug"
      #data: Illuminate\Support\Collection {#3569 ▼
        #items: array:7 [▼
          "blueprint" => "blue"
          "title" => "title"
          "updated_by" => "433a4d7e-e814-40b9-8cad-f817548d218a"
          "updated_at" => 1634663531
          "name" => array:1 [▼
            0 => "4e88b58e-3276-485a-b423-28a40e319183"
          ]
          "my-data" => array:1 [▼
            0 => array:7 [▼
              "bard" => array:1 [▼
                0 => array:2 [▼
                  "type" => "paragraph"
                  "content" => array:1 [▶]
                ]
              ]
              "enabled" => true
              "links" => array:2 [▶]
            ]
          ]
          "name" => array:1 [▶]
        ]
      }
      #supplements: Illuminate\Support\Collection {#3582 ▶}
      #published: true
      #selectedQueryColumns: null
      #origin: null
    }
    1 => Statamic\Entries\Entry {#3635 ▶}
    2 => Statamic\Entries\Entry {#3633 ▶}
    3 => Statamic\Entries\Entry {#3661 ▶}
  ]
}
```

My structure looks the same. I even managed to wrap it into an Entry:
```
 $wrapped = $entries->map(function ($item, $key) {
            return Entry::make()
                ->collection("somename")
                ->slug("somename")
                ->data($my-data);
        });

```

Still no difference. But I think I found the main difference when i dump the values in the antlers template.
The working one, if i DO NOT use my custom tag will properly type each field like this:
```
array:2 [▼
  0 => array:5 [▼
    "bard_field_name" => Statamic\Fields\Value {#3362 ▼
      #raw: array:1 [▶]
      #handle: "bard_field_name"
      #fieldtype: Statamic\Fieldtypes\Bard {#3151 ▶}
      #augmentable: Statamic\Entries\Entry {#2963 ▶}
      #shallow: false
    }
```
I cannot even dump the values of my custom tag. It throws this error:
LOG.debug: Cannot render an array variable as a string: {{ bard_field_name }}

I also tried to extend not the `Statamic\Tags\Tags;`but rather the collection `Statamic\Tags\Collection\Collection;`hoping to inherit all this functionality. But somehow it's just not working.

My guess is that I'm breaking somehow the functionality when im cutting out field of the data, not returning the entry but rather subfields.

```
 $termine = Entry::query()
            ->where("collection", "col-name")
            ->get()
            ->flatMap(function ($entry) {
                $data-i-want = $entry->get("fieldname");
                for ($i = 0; $i < count($termine); $i++) {
                    $data-i-want[$i]["parent_title"] = $entry->get("title");
                }
                return  $data-i-want;
            })
            ->sortBy("datum")
            ->where("datum", ">=", date("Y-m-d"));
```

**Comment by hanneshoepfnerio** [comment]
Hi, i rebuild a test scenario and went through it line by line and I found a fix – but I'm missing explanation or a more straight forward approach.

The flatMap did break it. map() was also enough for what i needed.
`   ->flatMap(function ($entry) {`--> `   ->map(function ($entry) {`

I wrap it as Entry but I needed to set the correct collection and field attribute, without it its not working:
```
 $test = collect($entries)->map(function ($item, $key) {
            return Entry::make()
                ->collection("the-right-collection-name")
                ->data(["the-original-field-name" => $item]);
        });
```
I'm assuming that it can resolve the bard field, since the information how to resolve fieldtypes is related to the collection (and most properly their blueprint).

I'm still curious if there is a better way to do this.
Thanks
